File: cpp/structure/share.cpp
#include <iostream>
#include <map>
using namespace std;

class Share {
};
class ShareFac {
private:
    static map<int, Share*> smap;
public:
    static Share* get(int i) {
        Share* s = smap[i];
        if (s == NULL) {
            s = new Share();
            smap[i] = s;
        }
        return s;
    }
};
map<int, Share*> ShareFac::smap;
int main() {
    Share* s1 = ShareFac::get(1);
    Share* s2 = ShareFac::get(1);
    Share* s3 = ShareFac::get(2);
    cout<<(s1 == s2) <<endl;
    cout<<(s1 == s3) <<endl;
}

File: golang/structure/share.go
package main

import "fmt"

type Share struct {
	a int
}

var smap = make(map[int] *Share)
func init() {
	println("init")
	smap[1] = &Share{}
	smap[2] = &Share{}
	fmt.Println(smap)
}
func  get(id int) *Share{
	fmt.Println(smap)
	return smap[id]
}

func main() {
	println("main")
	s1 := get(1)
	s2 := get(1)
	s3 := get(2)
	println(s1)
	println(s2)
	println(s3)
}